qiskit/transpiler/target.py

```diff
@@ -0,0 +1,652 @@
+        :class:`~qiskit.transpiler.InstructionProperties` object. This pattern
+        is repeated for each :class:`~qiskit.circuit.Instruction` the target
+        supports.
+
+        Args:
+            instruction (Instruction): The gate object to add to the map. If it's
+                paramerterized any value of the parameter can be set
+            properties (dict): A dictionary of qarg entries to an
+                :class:`~qiskit.transpiler.InstructionProperties` object for that
+                instruction implementation on the backend. Properties are optional
+                for any instruction implementation, if there are no
+                :class:`~qiskit.transpiler.InstructionProperties` available for the
+                backend the value can be None. If there are no constraints on the
+                instruction (as in a noisless/ideal simulation) this can be set to
+                ``{None, None}`` which will indicate it runs on all qubits (or all
+                available permutations of qubits for multi-qubit gates). The first
+                ``None`` indicates it applies to all qubits and the second ``None``
+                indicates there are no
+                :class:`~qiskit.transpiler.InstructionProperties` for the
+                instruction. By default, if properties is not set it is equivalent to
+                passing ``{None: None}``.
+            name (str): An optional name to use for identifying the instruction. If not
+                specified the :attr:`~qiskit.circuit.Instruction.name` attribute
+                of ``gate`` will be used. All gates in the ``Target`` need unique
+                names. Using a custom name allows a backend to specify duplicate
+                gates with different parameters.
+        Raises:
+            AttributeError: If gate is already in map
+        """
+        if properties is None:
+            properties = {None: None}
+        instruction_name = name or instruction.name
+        if instruction_name in self._gate_map:
+            raise AttributeError("Instruction %s is already in the target" % instruction_name)
+        self._gate_name_map[instruction_name] = instruction
+        qargs_val = {}
+        for qarg in properties:
+            if qarg is not None:
+                self.num_qubits = max(self.num_qubits, max(qarg) + 1)
+            qargs_val[qarg] = properties[qarg]
+            self._qarg_gate_map[qarg].add(instruction_name)
+        self._gate_map[instruction_name] = qargs_val
+        self._coupling_graph = None
+        self._instruction_durations = None
+        self._instruction_schedule_map = None
+
+    def update_instruction_properties(self, instruction, qarg, properties):
+        """Update the property object for an instruction qarg pair already in the Target
+
+        Args:
+            instruction (str): The instruction name to update
+            qarg (tuple): The qarg to update the properties of
```

Comment:
Should this be `qargs`, since `Instruction` will in general be have more than one `qarg`?

Reply:
I was trying to make this the method to update the properties payload of a single argument, so to update them all you'd have to call this multiple times. Are you advocating for making this a list of arguments and updating multiple `InstructionProperties` payloads at once. If so how do you expect people to pass the separate payloads for each argument?

Reply:
No, I don't think so, though I do think maybe I'm getting a bit tripped up on the terminology. For example, in updating the `InstructionProperties` for a single `Instruction`, say a `CXGate` on qubits `3` and `7`, is the `(3, 7)` tuple a single `qarg`, or the `Instruction`'s `qargs`? I would've expected the latter, but it looks like here we're using `qargs` to refer to collections of these tuples.

Reply:
That is how I was using it here qarg is 1 tuple and qargs was >1 tuple. I don't think we're really consistent about things throughout qiskit. I can change things I guess, it's just adding an `s` here. I'll defer to whatever you think is clearer.

Reply:
Agree, I think that'd be better since it'd be good to be consistent here (and elsewhere) with the naming we use in `QuantumCircuit.append`.

Reply:
Done in: https://github.com/Qiskit/qiskit-terra/pull/5885/commits/599d2fda04ea02a54bba035b5aeaf82e1a123c2a